### ISSUE TITLE
Rule Proposal: no-empty-pattern [$10]

### ISSUE THREAD

**Issue by mysticatea** [eslint/eslint#3668]
Spawned from #3658.

The following code does not raise an syntax error.

```js
let {a: {}} = foo;
```

But it's considered developer's intent is below.
Because an empty `ObjectPattern` effects nothing.

```js
let {a = {}} = foo;
```

So this proposal defines new `no-empty-pattern` rule.
This rule considers the following patterns as warnings:

```js
let {} = foo;
let [] = foo;
let {a: {}} = foo;
let {a: []} = foo;
function foo({}) {}
function foo([]) {}
function foo({a: {}}) {}
function foo({a: []}) {}
```

This rule does not consider the following patterns as warnings:

```js
let {a = {}} = foo;
let {a = []} = foo;
function foo({a = {}}) {}
function foo({a = []}) {}
```

----
I'm not sure this rule name is good.

**Comment by mysticatea** [comment]
`import {} from "foo";` is the similar one.

**Comment by mysticatea** [comment]
Is `no-empty-declaration` a good name?
These patterns does not declare any variable.

**Comment by ilyavolodin** [comment]
What about `var` and `const`? I assume those are going to be affected too?

**Comment by mysticatea** [comment]
Yes, this rule covers `var` and `const`.

I thought this rule is super simple.

```js
return {
    "ObjectPattern": function(node) {
        if (node.properties.length === 0) {
            context.report(node, "Unexpected empty object pattern.");
        }
    },
    "ArrayPattern": function(node) {
        if (node.elements.length === 0) {
            context.report(node, "Unexpected empty array pattern.");
        }
    }
};
```

**Comment by nzakas** [comment]
Yeah, seems like a good idea

**Comment by alberto** [comment]
@mysticatea are you working on this?

**Comment by mysticatea** [comment]
@alberto No yet. I am happy if you work on this :)

**Comment by alberto** [comment]
Thanks, I'll give it a try. :)

**Comment by nzakas** [comment]
@alberto there's a bounty on this issue that you can claim: https://www.bountysource.com/issues/26471568-rule-proposal-no-empty-pattern?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github

**Comment by alberto** [comment]
That should go to @mysticatea, it was all his code. :)

**Comment by nzakas** [comment]
Okay, @mysticatea there's money for you!
